test/xskbench/km/sys/driver.c

```diff
@@ -30,205 +25,73 @@ static BOOLEAN IsSessionActive;
-    UNREFERENCED_PARAMETER(ClientBindingContext);
-
-    TraceEnter(TRACE_CONTROL, "-");
-
-    //
-    // Initiate the closure of all XDPAPI handles.
-    //
-
-    // return STATUS_PENDING;
-
-    XdpApiProviderBindingContext = NULL;
-    XdpApi = NULL;
-    KeResetEvent(&BoundToProvider);
-
-    TraceExitSuccess(TRACE_CONTROL);
-
-    return STATUS_SUCCESS;
-}
-
-// VOID
-// XbDrvAllXdpApiHandlesAreClosed(
-//     XBDRV_NMR_CLIENT_BINDING_CONTEXT* BindingContext
-//     )
-// {
-//     //
-//     // Indicate detach completion.
-//     //
-//     NmrClientDetachProviderComplete(BindingContext->NmrBindingHandle);
-// }
-
 VOID
-XbDrvNmrCleanupXdpApiBindingContext(
-    PVOID ClientBindingContext
+XdpDetach(
+    _In_ VOID *ClientContext
     )
 {
-    XBDRV_NMR_CLIENT_BINDING_CONTEXT* BindingContext =
-        (XBDRV_NMR_CLIENT_BINDING_CONTEXT*)ClientBindingContext;
-
-    TraceEnter(TRACE_CONTROL, "-");
-
-    //
-    // Free memory for this binding.
-    //
-    ExFreePool(BindingContext);
-
-    TraceExitSuccess(TRACE_CONTROL);
+    UNREFERENCED_PARAMETER(ClientContext);
 }
 
-const NPI_CLIENT_CHARACTERISTICS XbDrvNmrXdpApiClientCharacteristics = {
-    0, // Version
-    sizeof(NPI_CLIENT_CHARACTERISTICS),
-    (PNPI_CLIENT_ATTACH_PROVIDER_FN)XbDrvNmrAttachXdpApiProvider,
-    (PNPI_CLIENT_DETACH_PROVIDER_FN)XbDrvNmrDetachXdpApiProvider,
-    (PNPI_CLIENT_CLEANUP_BINDING_CONTEXT_FN)XbDrvNmrCleanupXdpApiBindingContext,
-    {
-        0, // Version
-        sizeof(NPI_REGISTRATION_INSTANCE),
-        &NPI_XDPAPI_INTERFACE_ID,
-        &NPI_XBDRV_MODULEID,
-        XDP_API_VERSION_1, // Number
-        NULL // NpiSpecificCharacteristics
-    } // ClientRegistrationInstance
-};
-
-XDP_API_PROVIDER_BINDING_CONTEXT *
-CxPlatXdpApiGetProviderBindingContext(
-    VOID
-    )
-{
-    return XdpApiProviderBindingContext;
-}
+static XDP_API_CLIENT XdpApiContext = {0};
+#define TEST_TIMEOUT_ASYNC_MS 1000
+#define POLL_INTERVAL_MS 10
 
 VOID
 CxPlatXdpApiInitialize(
     VOID
     )
 {
     NTSTATUS Status;
+    KEVENT Event;
+    INT32 TimeoutMs = TEST_TIMEOUT_ASYNC_MS;
 
-    TraceEnter(TRACE_CONTROL, "-");
-
-    Status =
-        NmrRegisterClient(
-            &XbDrvNmrXdpApiClientCharacteristics,
-            NULL,
-            &NmrRegistrationHandle);
+    Status = XdpLoadApi(XDP_API_VERSION_1, NULL, NULL, &XdpDetach, &XdpFuncXdpApiClientDispatch, &XdpApiContext);
     if (!NT_SUCCESS(Status)) {
         goto Done;
     }
 
-    KeWaitForSingleObject(&BoundToProvider, Executive, KernelMode, FALSE, NULL);
+    KeInitializeEvent(&Event, NotificationEvent, FALSE);
+
+    do {
+        LARGE_INTEGER Timeout100Ns;
+
+        Status =
+            XdpOpenApi(
+                &XdpApiContext,
+                &XdpApi,
+                &XdpApiProviderBindingContext);
+        if (NT_SUCCESS(Status)) {
+            break;
+        }
+
+        Timeout100Ns.QuadPart = -1 * Int32x32To64(POLL_INTERVAL_MS, 10000);
+        KeResetEvent(&Event);
```

Comment:
Where does this event get set?

Reply:
Looks like this is for sleep before retrying